core/src/sumeragi/main_loop.rs

```diff
@@ -420,6 +431,50 @@ fn commit_block<F: FaultInjection>(
     cache_transaction(state, sumeragi);
 }
 
+fn replace_top_block<F: FaultInjection>(
+    sumeragi: &SumeragiWithFault<F>,
+    state: &mut State,
+    block: impl Into<VersionedCommittedBlock>,
+) {
+    let committed_block = block.into();
+    let block_hash = committed_block.hash();
+
+    state.wsv = state.finalized_wsv.clone();
+    state
+        .wsv
+        .apply(&committed_block)
+        .expect("Failed to apply block on WSV. This is absolutely not acceptable.");
+
+    sumeragi.send_events(state.wsv.events_buffer.replace(Vec::new()));
+    sumeragi.send_events(&committed_block);
+
+    // state.previous_block_hash stays the same.
+    state.latest_block_height = committed_block.header().height;
+    state.latest_block_hash = Some(block_hash);
+    state.latest_block_view_change_index = committed_block.header().view_change_index;
+
+    // Update WSV copy that is public facing
+    {
+        let mut wsv = sumeragi.wsv.lock();
+        *wsv = state.wsv.clone();
+    }
```

Comment:
```suggestion
        *sumeragi.wsv.lock() = state.wsv.clone();
```
Rust drops temporaries (values not bound to a name) immediately.